### PR TITLE
Add the Initial Version of Interactive Demo

### PR DESCRIPTION
**Pull request by dalgu90**
Hey team. I added an initial version of the interactive ICD coding demo.   
The demo app is written by Streamlit and can run only one model for now. We can run the app (a dataset and a model needed) by running a command like
```
$ streamlit run app.py -- --config_path configs/demo/caml_mimic3_50.yml
```
While implementing the app, I also updated other things in the code:
- Refactored the `base_dataset` class exposing encoding/decoding methods (of both input text and label). 
- Changed the type of error when the key misses in the `__getattr__()` of configuration (`KeyError` -> `AttributeError`).
- Changed the checkpoint saver not to change the input config.

Here's a screenshot of the demo.
image
Please nitpick any details of the code.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    disabled=True,
+)
+
+# Tokenize text with vocab
+token_idxs = dataset.encode_tokens(preprocessed_text.split())
+token_text = " ".join(dataset.decode_tokens(token_idxs))
+st.text_area(label="Tokens", value=token_text, height=200, disabled=True)
+
+# Model prediction
+st.write("Output")
```
Comment (abheesht17) [inline app.py:110]: Should we use `st.header()` here? Even a large font size works. What do you think?

File: app.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+from src.modules.embeddings import *
+from src.modules.preprocessors import ClinicalNotePreprocessor
+from src.utils.checkpoint_savers import *
+from src.utils.configuration import Config
+from src.utils.mapper import ConfigMapper
+
+hash_funcs = {
+    Config: lambda x: hash(str(x)),
+    torch.nn.parameter.Parameter: lambda x: hash(x.shape),
+}
```
Comment (abheesht17) [inline app.py:24]: Nice find! I did not know about this!

Comment (abheesht17) [inline app.py:24]: Nice find!

File: app.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    args = parser.parse_args()
+    config = Config(path=args.config_path)
+    return config
+
+
+@st.cache(hash_funcs=hash_funcs, allow_output_mutation=True)
+def load_modules(config):
+    # Load preprocessor
+    preprocessor = ClinicalNotePreprocessor(config.clinical_note_preprocessing)
+    preprocessor("")
```
Comment (abheesht17) [inline app.py:45]: Hey, what does this line do? Is this some sort of placeholder call?

Comment (abheesht17) [inline app.py:45]: Hey, what does this line do? Is this some sort of placeholder call?

**Review by abheesht17 — COMMENTED** [review]
This looks great, @dalgu90! Thank you for the PR. Some very minor comments.

As discussed during our call earlier this week, I guess there are a few to-dos, which we can do in a separate PR:
- Making a description expander for every model.
- Keeping a sidebar [needs more discussion].
- Using Integrated Gradients/Gradient Attribution.

We can split this work among ourselves if you want. :)

File: app.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    top_k_codes = dataset.decode_labels(top_k_preds)
+    top_k_descs = [icd_desc[c] for c in top_k_codes]
+
+    # Output as table
+    output_df = pd.DataFrame(
+        {
+            "ICD-9 Code": top_k_codes,
+            "Probability": top_k_probs,
+            "Description": top_k_descs,
+        }
```
Comment (abheesht17) [inline app.py:130]: Random thought. Here, can we have something like what is given in the image?
image

Currently, we have this: 
image

I've heard people use Plotly tables. The options in Streamlit are st.table, st.dataframe and Plotly tables [https://discuss.streamlit.io/t/exploring-the-limits-of-dataframes-and-tables-in-streamlit/879].

Can we have tables in Streamlit with expandable rows, etc.?

File: app.py (hunk 3 above)
Comment (dalgu90) [reply app.py:45]: Thanks for the question. That line (L44) was written for some purpose, but now it can be removed.  
The line is called to load all the inner modules in the ClinicalNotePreprocessor (some inner modules seem to be loaded lazily), so the returned `preprocessor` will not be changed later. But now that we have `allow_output_mutation=True` in the `st.cache()` annotator, it's okay for the `preprocessor` can be changed now.

File: app.py (hunk 1 above)
Comment (dalgu90) [reply app.py:110]: Thanks for the comment. That small title "Output" is intended and will be changed in a later version with other titles.  
Currently, we are using `st.text_area()` for the input text forms, and the option `label` is a required parameter to show a small title on each form. So, the small "output" title on the model prediction goes well with these labels.

File: app.py (hunk 4 above)
Comment (dalgu90) [reply app.py:130]: Thanks for the comment. If it's the problem for the descriptions not showing in full. We can change it simply to `st.table()` instead of `st.dataframe()`. It can be outputted like this (the descriptions are repeated 5 times to make them longer):
image
Let me take a look at the Plotly option but for now I will change it to `st.table()`.

Comment (abheesht17) [reply app.py:129]: Yep, this looks great! Thanks!

File: app.py (hunk 3 above)
Comment (abheesht17) [reply app.py]: Ah, I see. Thanks for the detailed explanation!

**Review by abheesht17 — APPROVED** [review]
LGTM! Thank you for addressing the comments. Could you please run the pre-commit hooks (if you haven't run them already)?

**Comment by dalgu90** [comment]
Thanks for the approval @abheesht17 . I just passed the pre-commit hooks and will merge this PR.
